src/bgw/job.c

```diff
@@ -67,15 +71,132 @@ ts_bgw_job_start(BgwJob *job, Oid user_oid)
 	return ts_bgw_start_worker(NameStr(job->fd.application_name), &bgw_params);
 }
 
+static void
+job_execute_function(FuncExpr *funcexpr)
+{
+	bool isnull;
+
+	EState *estate = CreateExecutorState();
+	ExprContext *econtext = CreateExprContext(estate);
+
+	ExprState *es = ExecPrepareExpr((Expr *) funcexpr, estate);
+	ExecEvalExpr(es, econtext, &isnull);
+	FreeExprContext(econtext, true);
+	FreeExecutorState(estate);
+}
+
+/* This is copied from tsl/src/job.c */
+static void
+job_execute_procedure(FuncExpr *funcexpr)
+{
+	CallStmt *call = makeNode(CallStmt);
+	call->funcexpr = funcexpr;
+	DestReceiver *dest = CreateDestReceiver(DestNone);
+	/* we don't need to create proper param list cause we pass in all arguments as Const */
+#ifdef PG11
+	ParamListInfo params = palloc0(offsetof(ParamListInfoData, params));
+#else
+	ParamListInfo params = makeParamList(0);
+#endif
+	ExecuteCallStmt(call, params, false, dest);
+}
+
+/**
+ * Run configuration check validation function.
+ *
+ * This will run the configuration check validation function registered for
+ * the job. If a new job is added, the job_id is going to be zero.
+ */
+void
+ts_bgw_job_run_config_check(Oid check, int32 job_id, Jsonb *config)
+{
+	/* Nothing to check if there is no check function provided */
+	if (!OidIsValid(check))
+		return;
+
+	List *args =
+		list_make1(makeConst(JSONBOID, -1, InvalidOid, -1, JsonbPGetDatum(config), false, false));
+	FuncExpr *funcexpr =
+		makeFuncExpr(check, VOIDOID, args, InvalidOid, InvalidOid, COERCE_EXPLICIT_CALL);
+
+	switch (get_func_prokind(check))
+	{
+		case PROKIND_FUNCTION:
+			job_execute_function(funcexpr);
+			break;
+		case PROKIND_PROCEDURE:
+			job_execute_procedure(funcexpr);
+			break;
+		default:
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED), errmsg("unsupported function type")));
+			break;
+	}
+}
+
+/* Run the check function on a configuration. It will generate errors if there
+ * is anything wrong with the configuration, otherwise just return. If the
+ * check function does not exist, no checking will be done.*/
+static void
+job_config_check(BgwJob *job, Jsonb *config)
+{
+	Oid proc;
+	ObjectWithArgs *object;
+	bool started = false;
+
+	/* Both should either be empty or contain a schema and name */
+	Assert((strlen(NameStr(job->fd.check_schema)) == 0) ==
+		   (strlen(NameStr(job->fd.check_schema)) == 0));
+
+	/* If there is no function, just return */
+	if (strlen(NameStr(job->fd.check_name)) == 0)
+		return;
+
+	if (!IsTransactionOrTransactionBlock())
+	{
+		started = true;
+		StartTransactionCommand();
+		/* executing sql functions requires snapshot */
+		PushActiveSnapshot(GetTransactionSnapshot());
```

Comment:
Why do we need that here? Do we need to create a portal like in `tsl/src/bgw_policy/job.c`?

```
#if (PG12 && PG_VERSION_NUM >= 120008) || (PG13 && PG_VERSION_NUM >= 130004) || PG14_GE
		EnsurePortalSnapshotExists();
#else
```

Reply:
Oh yes I think you are right and this must be updated accordingly. (I say think because I didn't add these particular lines, but the logic should be the same as in `job_execute`). Which reminds me - I didn't test what happens if the check function contains a `COMMIT` .

Reply:
Can the "check function" be a PROCEDURE? I guess no because procedures doesn't return values and as far I understood the check function should return true or false, right?

Reply:
Would be nice add some tests for it to make sure we'll not introduce regressions

Reply:
Thank you for pointing out this part of the code. I think that we do not need that here after all. The if condition is essentially never satisfied, since it would only be true for a background worker in which no transaction has been started yet. We also don’t need a portal. The portal was introduced in #3574 to allow procedures containing COMMIT statements to be executed by a background worker. Our add_job and alter_job are functions, not procedures, so the check function is not allowed to do any transaction control. (But that would not be its purpose either). 
I added a test - error case for this instance.